src/NATS.Client.JetStream/NatsJSOrderedConsumer.cs

```diff
@@ -180,23 +189,65 @@ public async IAsyncEnumerable<NatsJSMsg<T>> FetchAsync<T>(
         [EnumeratorCancellation] CancellationToken cancellationToken = default)
     {
         cancellationToken = CancellationTokenSource.CreateLinkedTokenSource(_cancellationToken, cancellationToken).Token;
+        var processed = 0;
+        var bytesProcessed = 0;
 
-        var consumer = await RecreateConsumer(_fetchConsumerName, _fetchSeq, cancellationToken);
-        _fetchConsumerName = consumer.Info.Name;
-
-        await foreach (var msg in consumer.FetchAsync(opts, serializer, cancellationToken))
+        var retry = 0;
+        while (!cancellationToken.IsCancellationRequested)
         {
-            if (msg.Metadata is not { } metadata)
-                continue;
+            if ((opts.MaxMsgs.HasValue && processed >= opts.MaxMsgs) || (opts.MaxBytes.HasValue && bytesProcessed >= opts.MaxBytes))
+                yield break;
 
-            _fetchSeq = metadata.Sequence.Stream;
-            yield return msg;
-        }
+            var mismatch = false;
+            ulong cseq = 0;
+
+            var consumer = await RecreateConsumer(_fetchConsumerName, _fetchSeq, cancellationToken);
+            _fetchConsumerName = consumer.Info.Name;
+
+            try
+            {
+                var fetchOpts = opts with { MaxMsgs = opts.MaxMsgs - processed, MaxBytes = opts.MaxBytes - bytesProcessed };
+
+                await foreach (var msg in consumer.FetchAsync(fetchOpts, serializer, cancellationToken))
+                {
+                    if (msg.Metadata is not { } metadata)
+                        continue;
+
+                    var expected = cseq + 1;
+                    if (metadata.Sequence.Consumer != expected)
+                    {
+                        _logger.LogWarning(NatsJSLogEvents.Retry, "Consumer sequence mismatch. Expected {Expected}, was {SequenceConsumer}. Retrying...", expected, metadata.Sequence.Consumer);
+                        mismatch = true;
+                        break;
+                    }
+
+                    _fetchSeq = metadata.Sequence.Stream;
+                    cseq = metadata.Sequence.Consumer;
+
+                    processed++;
+                    bytesProcessed += msg.Size;
+
+                    yield return msg;
+                }
+            }
+            finally
+            {
+                var deleted = await TryDeleteConsumer(_fetchConsumerName, cancellationToken);
```

Comment:
Why do you care if the delete succeeds? It should be ephemeral anyway. The key is to always generate an ordered consumer name. We agreed that the user can provide a prefix for ordered consumer names, which will allow them to be found more easily in logs